libkbfs/interfaces.go

```diff
@@ -1447,3 +1447,177 @@ type RekeyQueue interface {
 	// Waits for all queued rekeys to finish
 	Wait(ctx context.Context) error
 }
+
+// ImmutableBareRootMetadata is a read-only interface to the bare serializeable MD that
+// is signed by the reader or writer.
+type ImmutableBareRootMetadata interface {
+	// TlfID returns the ID of the TLF this BareRootMetadata is for.
+	TlfID() TlfID
+	// LatestKeyGeneration returns the most recent key generation in this
+	// BareRootMetadata, or PublicKeyGen if this TLF is public.
+	LatestKeyGeneration() KeyGen
+	// IsValidRekeyRequest returns true if the current block is a simple rekey wrt
+	// the passed block.
+	IsValidRekeyRequest(codec Codec, prevMd ImmutableBareRootMetadata, user keybase1.UID) (
+		bool, error)
+	// MergedStatus returns the status of this update -- has it been
+	// merged into the main folder or not?
+	MergedStatus() MergeStatus
+	// IsRekeySet returns true if the rekey bit is set.
+	IsRekeySet() bool
+	// IsWriterMetadataCopiedSet returns true if the bit is set indicating
+	// the writer metadata was copied.
+	IsWriterMetadataCopiedSet() bool
+	// IsFinal returns true if this is the last metadata block for a given
+	// folder.  This is only expected to be set for folder resets.
+	IsFinal() bool
+	// IsWriter returns whether or not the user+device is an authorized writer.
+	IsWriter(user keybase1.UID, deviceKID keybase1.KID) bool
+	// IsReader returns whether or not the user+device is an authorized reader.
+	IsReader(user keybase1.UID, deviceKID keybase1.KID) bool
+	// DeepCopy returns a deep copy of the underlying data structure.
+	DeepCopy(codec Codec) (BareRootMetadata, error)
+	// CheckValidSuccessor makes sure the given BareRootMetadata is a valid
+	// successor to the current one, and returns an error otherwise.
+	CheckValidSuccessor(currID MdID, nextMd ImmutableBareRootMetadata) error
+	// CheckValidSuccessorForServer is like CheckValidSuccessor but with
+	// server-specific error messages.
+	CheckValidSuccessorForServer(currID MdID, nextMd ImmutableBareRootMetadata) error
+	// MakeBareTlfHandle makes a BareTlfHandle for this
+	// BareRootMetadata. Should be used only by servers and MDOps.
+	MakeBareTlfHandle() (BareTlfHandle, error)
+	// VerifyWriterMetadata verifies md's WriterMetadata against md's
+	// WriterMetadataSigInfo, assuming the verifying key there is valid.
+	VerifyWriterMetadata(codec Codec, crypto cryptoPure) error
+	// TlfHandleExtensions returns a list of handle extensions associated with the TLf.
+	TlfHandleExtensions() (extensions []TlfHandleExtension)
+	// Returns the TLF key bundles for this metadata at the given key generation.
+	GetTLFKeyBundles(keyGen KeyGen) (*TLFWriterKeyBundle, *TLFReaderKeyBundle, error)
```

Comment:
this allows mutable access, so this should more properly be in the mutable interface